flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeIteratorWithCache.java

```diff
@@ -0,0 +1,367 @@
+        private final CountDownLatch readLatch;
+
+        private volatile File file;
+        private volatile FileOutputStream fo;
+        private volatile Exception saveCacheDataException;
+
+        FileCache(File cacheDir) {
+            this.cacheDir = cacheDir;
+            this.writeInBytes = new AtomicLong(0);
+            this.writeComplete = new AtomicBoolean(false);
+            this.refCount = new AtomicInteger(0);
+            this.readLatch = new CountDownLatch(1);
+        }
+
+        String getFilePath() {
+            return this.file.getAbsolutePath();
+        }
+
+        OutputStream getOutputStreamForSaveCacheData() throws IOException {
+            synchronized (this) {
+                if (fo == null) {
+                    file = File.createTempFile(CACHE_FILE_PREFIX, null, cacheDir);
+                    fo = new FileOutputStream(file);
+                    readLatch.countDown();
+                } else {
+                    throw new IllegalStateException("only can get OutputStream once !");
+                }
+            }
+
+            return new OutputStream() {
+                @Override
+                public void write(int b) throws IOException {
+                    fo.write(b);
+                    writeInBytes.incrementAndGet();
+                }
+
+                @Override
+                public void write(byte[] b, int off, int len) throws IOException {
+                    fo.write(b, off, len);
+                    writeInBytes.addAndGet(len);
+                }
+
+                @Override
+                public void close() throws IOException {
+                    fo.close();
+                    writeComplete.set(true);
+                }
+            };
+        }
+
+        void setSaveCacheDataException(Exception e) {
+            this.saveCacheDataException = e;
+        }
+
+        int getRefCount() {
+            return refCount.get();
+        }
+
+        private void handoverException() throws IOException {
+            if (saveCacheDataException != null) {
+                throw new IOException(
+                        "there is a exception when save data to cache file : ",
+                        saveCacheDataException);
+            }
+        }
+
+        FileInputStream open() throws IOException {
+            return open0();
+        }
+
+        FileInputStream openAndSeek(long offset) throws IOException {
+            FileInputStream fin = open0();
+            if (offset != 0) {
+                LOG.debug("seek to {}", offset);
+                fin.getChannel().position(offset);
+            }
+            return fin;
+        }
+
+        private FileInputStream open0() throws IOException {
+            try {
+                readLatch.await();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+
+            refCount.incrementAndGet();
+
+            return new FileInputStream(file) {
```

Comment:
This creates a new file descriptor and potentially calls `fopen` for the same file - for every offset in it.
Can we reuse the stream, at least for the same thread?
Or maybe it would be simpler to serialize all accesses to the same file in this PR? Seems like it will change the solution substantially, e.g. by alleviating all the concurrency in `FileCache`.

Reply:
I tend to open a new ticket to discuss serialize reads, WDYT @rkhachatryan  ？

Reply:
I think it makes sense @zoltar9264:
- the new implementation doesn't seem to be affected much by access serialization
- and there should be no performance regression after this PR

Reply:
Thanks @rkhachatryan , If there are any problems with the new implementation, just let me know and I will fix it as soon as possible.